Handle Windows CRT locale names

utelle opened the pull request:
This PR fixes issue https://github.com/wxWidgets/wxWidgets/issues/22247 by handling Windows CRT locale names properly.

Additionally, method `FindLanguageInfo` was fixed to support human readable locale strings like "English_United States" as claimed by the documentation. The previous implementation was broken. However, the separator between language and region has to be an underscore (not a hyphen). The interface description was updated to show "Portuguese_Brazil" (instead of "Portuguese-Brazil").

**Note**: The final return call to `DoPostInit` in method `wxLocale::Init((const wxString& name, const wxString& shortName = wxEmptyString, const wxString& locale = wxEmptyString, bool bLoadDefault = true);` now uses the parameter `strShort` (instead of `shortName`). `shortName` can be empty, and using an empty string for the language of the translation catalog doesn't make much sense, does it? The drawback of this is that `strShort` and `shortName` could possibly be different. However, this can only happen, if inconsistent parameters were specified like `locale="fr"` and `shortName="de"`.

utelle committed:
Handle Windows CRT locale names

utelle commented:
All Unix CI builds fail with

> /home/runner/work/wxWidgets/wxWidgets/tests/net/webrequest.cpp:152: FAILED:
> REQUIRE( stateFromEvent == requiredState )
> with expansion:
> 4 == 3

However, this seems to be unrelated to `wxLocale` and `wxUILocale`.

vadz commented:
The failures are due to https://github.com/chromium/badssl.com/issues/500 so I'm going to disable its use for now.

vadz commented on interface/wx/uilocale.h:
```diff
@@ -374,10 +374,11 @@ class wxLocaleIdent
-        - macOS:   \<language\>[-\<script\>][_\<region\>]
+        - BCP-47:      \<language\>[-\<script\>][-\<region\>][-\<extension\>]
+        - Windows:     \<language\>[-\<script\>][-\<region\>][-\<extension\>][_\<sortorder\>]
+        - Windows CRT: \<language\>[_\<region\>][.\<charset\>]
```
This seems to be a strict substring of the POSIX format just below, do we need to document it separately? I think it could be more useful to just write that region names are OS-dependent.

If we do document it, we should probably call it "MSVC" as other Windows compilers are perfectly free to use different formats.

utelle replied:
> This seems to be a strict substring of the POSIX format just below, do we need to document it separately? I think it could be more useful to just write that region names are OS-dependent.

The structure is the same, but "language" and "region" do not conform to ISO 639 resp ISO 3166 as is the case for POSIX. "language" and "region" are plain English names (at least I hope they are restricted to English). Not sure whether we need to document this here.

> If we do document it, we should probably call it "MSVC" as other Windows compilers are perfectly free to use different formats.

IMHO we should document it, because it's really not the same type of tag, even if the structure is the same. We can call it MSVC, although I find this too restrictive, because MinGW uses the same Microsoft runtime library. I don't know whether other compilers bring their own implementation.

vadz commented on interface/wx/intl.h:
```diff
@@ -400,7 +400,7 @@ class wxLocale
         given locale, specified either as a two letter ISO language code (for example,
         "pt"), a language code followed by the country code ("pt_BR") or a full, human
-        readable, language description ("Portuguese-Brazil").
+        readable, language description ("Portuguese_Brazil").
```
Does this mean "Portuguese-Brazil" doesn't and won't work? I think that ideally both it and the version with the underscore should work.

utelle replied:
> Does this mean "Portuguese-Brazil" doesn't and won't work? I think that ideally both it and the version with the underscore should work.

Right, "Portuguese-Brazil" won't work. And I don't intend to change that - at least at the moment, unless you convince me to do otherwise. The reason is simple: there are a few regions that have a hyphen in their name, for example "Portuguese_Guinea-Bissau" or "Portuguese_Timor-Leste". And theoretically the language name could also be a composed name with a hyphen (although we currently don't have such an entry in our language database).

If you insist to support "Portuguese-Brazil" as well, we could do so by first checking for an underscore separator, and - if not found - for a hyphen in a second step.

utelle replied:
In the light of the fact, that this form of specifying the language never worked in previous wx versions, it seems reasonable to restrict the syntax to an underscore separator. If there is a demand, this can be changed later. Nevertheless, a note about the allowed separator probably makes sense, because in the documentation of previous wx versions we have still the sample with a hyphen separator.

vadz commented on src/common/intl.cpp:
```diff
@@ -270,26 +273,35 @@ bool wxLocale::Init(const wxString& name,
-    if ( const wxLanguageInfo* langInfo = FindLanguageInfo(szLocale) )
+    int languageId = wxLANGUAGE_UNKNOWN;
+    wxLocaleIdent localeId = wxLocaleIdent::FromTag(szLocale);
+    if ( const wxLanguageInfo* langInfo = wxUILocale::FindLanguageInfo(localeId) )
```
Should we check for `!localeId.IsEmpty()` here or is it guaranteed that `FindLanguageInfo()` returns null for it?

utelle replied:
> Should we check for `!localeId.IsEmpty()` here or is it guaranteed that `FindLanguageInfo()` returns null for it?

Good question. I will check that, and will modify the code, if necessary.

vadz commented on src/common/intl.cpp:
```diff
@@ -270,26 +273,35 @@ bool wxLocale::Init(const wxString& name,
+        else
+        {
+            strName = langInfo->Description;
+            strShort = langInfo->CanonicalRef.empty() ? langInfo->CanonicalName : langInfo->CanonicalRef;
```
We probably should add `wxLanguageInfo::GetCanonicalRefOrName()` (except with a better name if you can come up with one) instead of writing the same not particularly clear `CanonicalRef.empty() ? CanonicalName : CanonicalRef` expression in several places.

utelle replied:
> We probably should add `wxLanguageInfo::GetCanonicalRefOrName()` (except with a better name if you can come up with one) instead of writing the same not particularly clear `CanonicalRef.empty() ? CanonicalName : CanonicalRef` expression in several places.

Yes, makes sense. At first glance, I thought `wxLanguageInfo::GetCanonicalWithRegion()` might be a good name, because usually that's what you get. However, there are several exceptions (mostly for exotic languages). Nevertheless, this method name seems to be more suitable.

vadz reviewed (commented):
Thanks for fixing this!

Please let me know if you plan to change/add anything in the light of the remarks below, otherwise I'll just merge it like this, maybe with some tests if I have time to add them.

Thanks again!

utelle committed:
Handle review feedback

- rename locale format "Windows CRT" to "MSVC CRT" and update documentation
- add note that only underscore is supported as separator between language and region in search parameter of `wxLocale::FindLanguageInfo`
- introduce helper method `wxLanguageInfo::GetCanonicalWithRegion()`

utelle commented:
> Thanks for fixing this!

You are welcome.

> Please let me know if you plan to change/add anything in the light of the remarks below, otherwise I'll just merge it like this, maybe with some tests if I have time to add them.

I think I addressed all remarks. From my side no further changes are planned at the moment.

vadz commented:
Thanks, I've (squash-) merged this now and added a few trivial tests which is still not enough, but better than nothing.

One surprising thing I noticed is that we have 2 completely different implementations for `wxUILocale::FindLanguageInfo()` overloads taking strings and `wxLocaleIdent`. Shouldn't it be possible to somehow merge/reuse them? Of course, this is not exactly urgent.

Thanks again for fixing this!

utelle commented:
> Thanks, I've (squash-) merged this now and added a few trivial tests which is still not enough, but better than nothing.

It will certainly make sense to add more tests. However, I think the most important things are already covered.

> One surprising thing I noticed is that we have 2 completely different implementations for `wxUILocale::FindLanguageInfo()` overloads taking strings and `wxLocaleIdent`. Shouldn't it be possible to somehow merge/reuse them? Of course, this is not exactly urgent.

Well, the search loops in both implementations follow the same principle, but the search conditions are different. For strings the canonical name and the description are inspected, while for `wxLocaleIdent` only the locale tag attribute is inspected. The former allows to search for human readable locale names (aka MSVC CRT locale names), the latter is quicker using only 2 instead of 4 string comparisons. The actual search procedure could be extracted into a separate method, parameterized with a search condition function.
